src/core/libraries/dialogs/ime_dialog.cpp

```diff
@@ -1,28 +1,80 @@
 // SPDX-FileCopyrightText: Copyright 2024 shadPS4 Emulator Project
 // SPDX-License-Identifier: GPL-2.0-or-later
 
+#include <algorithm>
+#include <magic_enum.hpp>
 #include "common/logging/log.h"
 #include "core/libraries/error_codes.h"
 #include "core/libraries/libs.h"
 #include "ime_dialog.h"
+#include "ime_dialog_ui.h"
+
+static constexpr float MAX_X_POSITIONS[] = {3840.0f, 1920.0f};
+static constexpr float MAX_Y_POSITIONS[] = {2160.0f, 1080.0f};
 
 namespace Libraries::ImeDialog {
 
-static OrbisImeDialogStatus g_ime_dlg_status = OrbisImeDialogStatus::ORBIS_IME_DIALOG_STATUS_NONE;
+static OrbisImeDialogStatus g_ime_dlg_status = OrbisImeDialogStatus::NONE;
+static OrbisImeDialogResult g_ime_dlg_result{};
+static ImeDialogState g_ime_dlg_state{};
+static ImeDialogUi g_ime_dlg_ui;
 
-int PS4_SYSV_ABI sceImeDialogAbort() {
-    LOG_ERROR(Lib_ImeDialog, "(STUBBED) called");
-    return ORBIS_OK;
+static bool IsValidOption(OrbisImeDialogOption option, OrbisImeType type) {
+    if (False(~option &
+              (OrbisImeDialogOption::MULTILINE | OrbisImeDialogOption::NO_AUTO_COMPLETION))) {
+        return false;
+    }
+
+    if (True(option & OrbisImeDialogOption::MULTILINE) && type != OrbisImeType::DEFAULT &&
+        type != OrbisImeType::BASIC_LATIN) {
+        return false;
+    }
+
+    if (True(option & OrbisImeDialogOption::NO_AUTO_COMPLETION) && type != OrbisImeType::NUMBER &&
+        type != OrbisImeType::BASIC_LATIN) {
+        return false;
+    }
+
+    return true;
 }
 
-int PS4_SYSV_ABI sceImeDialogForceClose() {
-    LOG_ERROR(Lib_ImeDialog, "(STUBBED) called");
-    return ORBIS_OK;
+static bool IsMemZero(const void* ptr, size_t size) {
+    return std::all_of(static_cast<const u8*>(ptr), static_cast<const u8*>(ptr) + size,
+                       [](u8 c) { return c == 0; });
 }
 
-int PS4_SYSV_ABI sceImeDialogForTestFunction() {
-    LOG_ERROR(Lib_ImeDialog, "(STUBBED) called");
-    return ORBIS_OK;
+Error PS4_SYSV_ABI sceImeDialogAbort() {
+    if (g_ime_dlg_status == OrbisImeDialogStatus::NONE) {
+        LOG_INFO(Lib_ImeDialog, "IME dialog not in use");
+        return Error::DIALOG_NOT_IN_USE;
+    }
+
+    if (g_ime_dlg_status != OrbisImeDialogStatus::RUNNING) {
+        LOG_INFO(Lib_ImeDialog, "IME dialog not running");
+        return Error::DIALOG_NOT_RUNNING;
+    }
+
+    g_ime_dlg_status = OrbisImeDialogStatus::FINISHED;
+    g_ime_dlg_result.endstatus = OrbisImeDialogEndStatus::ABORTED;
```

Comment:
An application should call sceImeDialogTerm after sceImeDialogAbort and resources are freed there. Also the application is not able to call sceImeDialogInit after sceImeDialogAbort because status is not NONE (it's FINISHED) so the UI will never be drawn twoice.